### PR TITLE
[server] Use image-builder from workspace cluster

### LINKED ISSUE #9248: Make server talk to image-builder in workspace cluster

### PR DESCRIPTION
**Pull request by csweichel**
## Description
This PR makes server use the image builder from workspace cluster.

**Note: ** merge and deploy this only once #9335 is merged, deployed and running.

## Related Issue(s)

Fixes #9248

## How to test
Start a workspace which requires an image-build

## Release Notes

```release-note
Use image-builder from workspace cluster
```

 - [x] /werft run with-vm=true

### CODE REVIEW

**Comment by kylos101** [comment]
Hi :wave: ,

@csweichel #9335 is deployed. :+1: :raised_hands: 

@geropl , do we have a governed workspace cluster in `staging` that is separate from webapp, so we can test similar to prod? If not, we may be able to do with am `ephemeral cluster`, or `workspace-preview`.

What scenarios do we have to test?
1. Test with an explicit image build, similar to gitpod.io/#imagebuild/<some context to image build> (it should image build)
2. Test a prebuild that needs an image build first (it should image build, before doing prebuild)
3. Test a fresh repo with a custom dockerfile by starting a regular workspace (it should image build)
4. Am I missing any other scenarios? :thinking: 

@aledbf is there anything we need to prepare for, that you can think of, that we should add [here](https://github.com/gitpod-io/gitpod/issues/7845)?

**Comment by geropl** [comment]
> do we have a governed workspace cluster in staging that is separate from webapp, so we can test similar to prod?

We don't have separate workspace clusters, for have EU/US as "full" installations, with the ws-manager being dynamically registered. So we should be able to test just like in prod. :+1:

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [inline components/server/src/workspace/gitpod-server-impl.ts:1558]: :+1:

**Comment by geropl** [comment]
BTW: reviewed yesterday, and code looks good! :pray: 

Now waiting for a preview to test and play around with.

**Comment by geropl** [comment]
@csweichel Can we move forward with this, or is it blocked somehow? :thinking: 
Also, happy to assign someone from WebApp for rebasing/testing etc. if that helps. :pray:

**Comment by csweichel** [comment]
@geropl I've rebased the branch - some additional testing would be handy :)

**Comment by aledbf** [comment]
> #9337 9335 is merged,

that is already up and running since gen42.

**Comment by geropl** [comment]
@jankeromnes Is looking into reviewing and testing this in preview env and staging. :+1:

File: components/ws-manager-api/typescript/src/client-provider.ts
Comment (jankeromnes) [inline components/ws-manager-api/typescript/src/client-provider.ts:95]: This must be unintentional, right?

```suggestion
            client = this.createConnection(WorkspaceManagerClient, info, grpcOptions);
```

**Review by jankeromnes — COMMENTED** [review]
Briefly skimmed the code, looks okay to me, will now do some testing.

Preview env seems down though:

/werft run

**Comment by jankeromnes** [comment]
Maybe Werft doesn't see review comments:

/werft run

**Comment by jankeromnes** [comment]
With the PR re-deployed, my first attempt failed with:

```
Request createWorkspace failed with message: 13 INTERNAL: grpc: failed to unmarshal the received message string field contains invalid UTF-8

Unknown Error: { "code": -32603 }
```

https://cw-imgbldr-mv-ii.preview.gitpod-dev.com/#https://github.com/jankeromnes/gitpod-build-error

Screenshot 2022-05-02 at 11 43 05


Bad deploy or fault in PR? 🤔

**Comment by csweichel** [comment]
@jankeromnes I don't really understand where this is coming from. It's most definitely something on the TypeScript/server side. When I talk to the image-builder through ws-manager using gpctl everything works as intended. Bumping grpc-js to the latest version had no effect.

Maybe we could have a joint debugging session tomorrow - I'd need some more in-depth Node/TypeScript knowledge here.

**Comment by jankeromnes** [comment]
Sure, would be happy to look into it together! I don't have much experience with gRPC, but I'd probably proceed with log statements in all the relevant code locations we expect to hit, then see where we fail to narrow it down. That, or we redeploy this PR with tracing and look at HoneyComb.

**Comment by csweichel** [comment]
/hold

We must merge and deplo https://github.com/gitpod-io/gitpod/pull/9714 before deploying this one

**Comment by csweichel** [comment]
@jankeromnes do you wanna have another go testing this PR? It should work now 🤞

**Comment by jankeromnes** [comment]
Sure! Thanks @csweichel -- will do more testing this afternoon 🤞

**Comment by jankeromnes** [comment]
Oh no, I was able to get one image build that failed as expected, but every other workspace I try to start (with or without Dockerfile) seems permanently stuck in Starting (in phases `initializing`, `unknown` or `pending`):

<table><tr><td>
Screenshot 2022-05-03 at 14 36 41
</td><td>
Screenshot 2022-05-03 at 14 35 18
</td></tr></table>

**Comment by geropl** [comment]
Put this into draft because people were confused what kind of "hold"-state this PR was in.
We're planning to get back to this right next week.

File: components/ws-manager-api/typescript/src/client-provider.ts
Comment (sagor999) [reply components/ws-manager-api/typescript/src/client-provider.ts:95]: @csweichel this is still not addressed.

**Review by sagor999 — APPROVED** [review]
lgtm, except for one minor typo, but approving with hold to address that fix.
/hold

Tested in preview env using @jankeromnes two test repos and everything worked as expected.

**Comment by sagor999** [comment]
/hold
we are holding this until we deploy XL cluster change. And after that we can try to get this PR reviewed and pushed.

**Comment by aledbf** [comment]
> we are holding this until we deploy XL cluster change. And after that we can try to get this PR reviewed and pushed.

We have that now in gen46. What's missing?

**Comment by csweichel** [comment]
We need to investigate if the new image builder log path is being used. I.e. the path where image builder reports the log location and server consumes it the same way we do for prebuilds.

**Comment by easyCZ** [comment]
I'm moving this into draft as currently it is unclear what remains to do on this PR. Feel free to bring back into review if this is in a reviewable (and mergeable) state.

**Comment by geropl** [comment]
> I'm moving this into draft as currently it is unclear what remains to do on this PR. Feel free to bring back into review if this is in a reviewable (and mergeable) state.

@easyCZ :+1: for moving into draft.

FYI, @kylos101 and I agreed that the next step is two things:
 - try if having workspaces run in workspace clusters actually does work (Team Workspace)
 - try and test if log streaming works from workspaces clusters (Team WebApp, #11024)

:information_source:  The [logs-PR](https://github.com/gitpod-io/gitpod/pull/11026) contains some of the changes from this branch, so I will rebase it once those one hit `main`.

**Comment by geropl** [comment]
To move forward I split up both aspects of this PR and put them behind feature flags. The 2nd PR is here: https://github.com/gitpod-io/gitpod/pull/11103 .

**Comment by sagor999** [comment]
@csweichel what would you like to do with this PR?

**Pull request closed without merging**
